### PR TITLE
Cache creature entity counts (#1207)

### PR DESCRIPTION
**Pull request by colinrgodsey**
**Cache ent counts (#1207)**
* Tested and works 👍 compared counts using both methods after a bunch of chunk loads/unload and whatnot.
* ~~Made it super generic... although this method is only used by SpawnerCreature.~~
* ~~Can also put the "a" method back how it was and just change SpawnerCreature. It does have an O(n) complexity based on the number of cached counts its keeping when adding and removing ents, so if we limit it to just SpawnerCreature we'll know it only keeps 4 cached counts.~~
* ~~Can also limit cache checks just to EntityCreature so we dont have to do anything special for items, xp orbs, and all that when adding or removing them to the world. ~~

~~**Water over lava (#1227)**~~
* ~~Works, just ported the patch over (exists in BlockFluids in 1.13)~~

### CODE REVIEW

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+ 
+-    public int a(Class<?> oclass) {
++    // Paper start - entity count cache
++    private int countEntityType(Class<?> oclass) {
+         int i = 0;
+         Iterator iterator = this.entityList.iterator();
+ 
+         while (iterator.hasNext()) {
+             Entity entity = (Entity) iterator.next();
+-            if (entity.shouldBeRemoved) continue; // Paper
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: Please restore this line

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
++                continue;
+             }
+ 
+             if (oclass.isAssignableFrom(entity.getClass())) {
+@@ -2464,6 +2494,52 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
++    public int getEntityCount(Class<?> oclass) { return a(oclass); } // Paper - OBFHELPER
++    public int a(Class<?> oclass) {
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: We need to avoid the method being virtual moved like this. maybe dont rename the method above ,copy the desired logic to a new method and make the original a(clazz) method immed short circuit to the new one.

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
++        if(countCache.containsKey(oclass)) {
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: containsKey is not needed here. use .get() directly and check for null

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
++        }
++
++        return false;
++    }
++
++    protected void updateEntityCount(Entity entity, boolean incr) {
++        if(entity == null) return;
++        if(shouldIgnoreForCount(entity)) return;
++
++        countCache.replaceAll((clazz, count) -> {
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: why replaceAll? we need to instead walk the super chain and insert by supers instead.

we can updateEntityCount(Class, boolean), then loop over the super chain until Entity is hit, then modify each.

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
++        if (entity instanceof EntityInsentient) {
++            EntityInsentient entityinsentient = (EntityInsentient) entity;
++            return entityinsentient.isTypeNotPersistent() && entityinsentient.isPersistent();
++        }
++
++        return false;
++    }
++
++    protected void updateEntityCount(Entity entity, boolean incr) {
++        if(entity == null) return;
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: please add whitespace before the ( on all these if statements too

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch (hunk 1 above)
Comment (colinrgodsey) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: so i actually removed this line because the count is being updated by the actual modifications to *World#entityList*, if i didnt count entities that are marked shouldBeRemoved, the count could end up off (at the end of the tick when its time to actually remove them from the list).

EDIT: And i did it this way, adding the functionality to the List itself, just because it seemed (and proved to be) rather error prone augmenting each place where entities were added and removed.

Comment (aikar) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: Then something has to be refactored here.
This boolean represents "This entity should not of been in this list in the first place"

You'll need to hook into where this is set to true/false at (at removeEntity, and the world add method) and hook count changes there too.

File: Spigot-Server-Patches/0334-Configurable-speed-for-water-flowing-over-lava.patch [hunk 6]
```diff
@@ -0,0 +1,65 @@
+From 35c565d267a562633607766e1264ac0cd4c01d87 Mon Sep 17 00:00:00 2001
```
Comment (aikar) [inline Spigot-Server-Patches/0334-Configurable-speed-for-water-flowing-over-lava.patch]: Please break this patch out into a separate PR so they can be reviewed independently.

There are issues with this patch (not your changes, the previous version of it was flawed) that we need to fix before bringing it back to 1.13

File: Spigot-Server-Patches/0334-Configurable-speed-for-water-flowing-over-lava.patch [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+From: Colin Godsey <crgodsey@gmail.com>
```
Comment (aikar) [inline Spigot-Server-Patches/0334-Configurable-speed-for-water-flowing-over-lava.patch]: Also, we need to keep the original author / timestamp of this patch.

https://github.com/PaperMC/Paper/blob/pre/1.13/removed/server/0017-Configurable-speed-for-water-flowing-over-lava.patch

**Comment by colinrgodsey** [comment]
ahhhh do not merge.... that was not the full fix

**Comment by colinrgodsey** [comment]
the cache updating is starting to feel a bit heavy... should I maybe limit the cache to only classes that extend IAnimal (the type bounds of EnumCreatureType)?

**Comment by aikar** [comment]
The original goal of this ticket was to mirror this logic: 
https://github.com/PaperMC/Paper/blob/76e1e4d79f15c025442fe71f9b91d56a2497e1b7/Spigot-Server-Patches/0048-Avoid-hopper-searches-if-there-are-no-items.patch

But at the world level...

simple int counts, just for the 3 "categories".

I think it would be ideal to keep it simple like this instead of all the complexity you are trying.

**Comment by colinrgodsey** [comment]
alright, got it reduced down and tested again

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                 }
+                 // CraftBukkit end */
+ 
++                /*
++                 * Paper - clear count cache every once in a while.
++                 * Cache can never be entirely trusted because
++                 * an entity may become persistent or not
++                 * during its lifetime.
++                 */
++                if (this.ticks % 600 == 0) worldserver.clearEntityCountCache();
```
Comment (colinrgodsey) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: Would like to draw attention to this too. Ultimately, we can't keep the cache up to date with 100% accuracy. An entity may become persistent or not at any time, so may not trigger the incr/decr appropriately. Given that this is just limited to the creature counts now, i think its ok if it drifts a bit (again, only under the scenario that something is adding an entity, *then* setting it persistent) and we just wipe the cache every once in a while.

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+ 
+                     if (k <= l1) {
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 004c3ec47..baf73a411 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -1,28 +1,20 @@
+ 	package net.minecraft.server;
+ 
+-import co.aikar.timings.Timings;
```
Comment (colinrgodsey) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: ... thanks intellij

Comment (aikar) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: You'll want to turn off auto optimizing imports while working on paper :)

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
+index f525fd1b4..494759a1c 100644
+--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
++++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
+@@ -113,7 +113,7 @@ public final class SpawnerCreature {
+                 // CraftBukkit end
+ 
+                 if ((!enumcreaturetype.c() || flag1) && (enumcreaturetype.c() || flag) && (!enumcreaturetype.d() || flag2)) {
+-                    k = worldserver.a(enumcreaturetype.a());
++                    k = worldserver.getCreatureCount(enumcreaturetype); // Paper - entity count cache
```
Comment (colinrgodsey) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch:50]: decided to just leave the original method alone

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
++            EntityInsentient entityinsentient = (EntityInsentient) entity;
++            if (entityinsentient.isTypeNotPersistent() && entityinsentient.isPersistent()) {
++                return;
++            }
++        }
++
++        Class<?> clazz = entity.getClass();
++
++        for (EnumCreatureType typ : EnumCreatureType.values()) {
++            if (typ.a().isAssignableFrom(clazz)) {
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: need an obfuscation helper here

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+@@ -2464,6 +2485,61 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
++    // Paper start - entity count cache
++    public int getCreatureCount(EnumCreatureType typ) {
++        Integer count = countCache.get(typ);
++
++        if (count == null) {
++            count = getEntityCount(typ.a());
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: need an obfuscation helper here

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
++    public int getEntityCount(Class<?> oclass) { return a(oclass); } // Paper - OBFHELPER
+     public int a(Class<?> oclass) {
+         int i = 0;
+         Iterator iterator = this.entityList.iterator();
+@@ -2464,6 +2485,61 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
++    // Paper start - entity count cache
++    public int getCreatureCount(EnumCreatureType typ) {
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: add the e please..

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+             guard();
++            if (o instanceof Entity) updateEntityCount((Entity)o, false); // Paper
+             return super.remove( o );
+         }
+ 
+@@ -82,6 +101,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+     public final Map<String, EntityHuman> playersByName = Maps.newHashMap(); // Paper - World EntityHuman Lookup Optimizations
+     public final List<Entity> k = Lists.newArrayList();
+     protected final IntHashMap<Entity> entitiesById = new IntHashMap();
++    private Map<EnumCreatureType, Integer> countCache = new EnumMap(EnumCreatureType.class); // Paper - entity count cache
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: move this down to be with the getCreatureCount method

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+-import java.util.Iterator;
+-import java.util.List;
+-import java.util.Random;
+-import java.util.UUID;
++
++import java.util.*;
+ 
+ import java.util.function.Function;
+ import java.util.function.Predicate;
+-import java.util.function.Supplier;
```
Comment (aikar) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: restore all the imports

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch (hunk 8 above)
Comment (aikar) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: well, entities don't really change persistence state often...
but spigot went years without worrying about that either.

I don't think we even need to auto clear
just clear in EntityInsentient the 2 place persistence is changed, and on breed case too.

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch (hunk 13 above)
Comment (colinrgodsey) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: woops, sorry used to scala (reserved term)

**Comment by colinrgodsey** [comment]
alright, got the fixes done and let it test overnight

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
+index 7b64ec27c..55606756e 100644
+--- a/src/main/java/net/minecraft/server/EntityInsentient.java
++++ b/src/main/java/net/minecraft/server/EntityInsentient.java
+@@ -42,6 +42,7 @@ public abstract class EntityInsentient extends EntityLiving {
+     public float[] dropChanceArmor;
+     // public boolean canPickUpLoot; // CraftBukkit - moved up to EntityLiving
+     public boolean persistent;
++    public boolean countsAgainstSpawnLimit = true; // Paper
```
Comment (colinrgodsey) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch:15]: added this to take care of the persistent inconsistencies. gets set to false based on whether its persistent when first "counted", and then thats used to figure out whether to decr when its removed

**Comment by colinrgodsey** [comment]
retest this please

**Comment by colinrgodsey** [comment]
@aikar think it looks good... travis maybe broke tho? seems to die on an earlier patch

**Review by aikar — APPROVED** [review]

File: Spigot-Server-Patches/0333-Add-entity-count-cache.patch [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
++            if (o instanceof Entity) updateEntityCount((Entity)o, false); // Paper
+             return super.remove( o );
+         }
+ 
+@@ -2464,6 +2497,53 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
++    // Paper start - entity count cache
++    private Map<EnumCreatureType, Integer> countCache = new EnumMap(EnumCreatureType.class);
```
Comment (stephan-gh) [inline Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: Not sure if this is worth it (depends on how often this is updated), but it might be more efficient to use an `int` array here and fill it based on `enum.ordinal()`. That avoids the `Integer` boxing and is pretty much what `EnumMap` does internally anyway.

Comment (colinrgodsey) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: good point, do hate the stupid boxing whenever you use generics

Comment (colinrgodsey) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: hmm so im working on another patch that could *also* use this optimization. was just going to make a static class in EnumCreatureType for this. There any good way to do that in just one patch and use the other as a dependency?

Comment (stephan-gh) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: Patches are applied sequentially, so if you apply one of the patches first, the others can access stuff the first patch has added. (I'm not sure how a static class in `EnumCreatureType` could be of any help here though...)

Comment (colinrgodsey) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: wouldnt be huge:

```java
static public class Counter {
        int[] counts = new int[EnumCreatureType.values().length];

        public void incr(EnumCreatureType type) {
            counts[type.ordinal()]++;
        }

        public void decr(EnumCreatureType type) {
            counts[type.ordinal()]--;
        }

        public int get(EnumCreatureType type) {
            return counts[type.ordinal()];
        }
    }
```

can take it or leave it tho 🤷‍♂️ ~~

EDIT Uhm, now that i look at it, i think you're right lol

Comment (stephan-gh) [reply Spigot-Server-Patches/0333-Add-entity-count-cache.patch]: That doesn't really look worth an extra class. ;) But in any case, if you need the creature count in another patch as well, why not just reuse the counts you already set up in this patch? :)
